### ISSUE TITLE
Introduce `TaskContainer.maybeNamed`

### ISSUE THREAD

**Issue by Tapchicoma** [gradle/gradle#8057]
Please, introduce new method `TaskContainer.maybeNamed(name: String): TaskProvider<Task>?` that will allow to get `TaskProvider<Task>?` without triggering actual task creation.

### Expected Behavior

`TaskContainer.maybeNamed(name: String): TaskProvider<Task>?` should return non-nullable `TaskProvider<Task>` if task with given `name` is already added using `TaskContainer.register(...)`, and should return `null` if there is no such task with given `name`.

Calling this method should not trigger actual task creation and configuration, and should be a part of [task configuration avoidance](https://docs.gradle.org/5.0/userguide/task_configuration_avoidance.html) api.

### Current Behavior

Currently, if I need to get and add lazy configuration to specific task without triggering creation of it, only one method is available: `TaskContainer.named(...): TaskProvider<Task>` that throws `UnknownDomainObjectException` if task with given name is not registered.

This leads to using either `try {} catch (e: UnknownDomainObjectException) {}` construction, or `configureEach {}`, or `matching {}`.

### Context
Currently introducing task configuration avoidance in Gradle plugin: plugin adds several "meta" tasks that trigger more specific tasks when it executed. To add dependency to such "meta" task, I need either get it or create, if it is not yet added.

### Your Environment
Gradle version: `4.10.3`

**Comment by JLLeitschuh** [comment]
Related to #6243

**Comment by big-guy** [comment]
> The main problem is that this tries to make an inherently asynchronous process (building project configuration) synchronous (return named task provider IF it's been registered at the current point in time).

A project is intended to be evaluated synchronously, so I don't think this is a concern.  Any concurrent evaluation today is a bug.

> I find that most of the time that I have to realize a task when I don't really want to is to capture dependsOn / finalizedBy semantics

Could you describe this a bit more?

@Tapchicoma 
> Currently introducing task configuration avoidance in Gradle plugin: plugin adds several "meta" tasks that trigger more specific tasks when it executed. To add dependency to such "meta" task, I need either get it or create, if it is not yet added.

Could you describe this a bit more?  There are a couple of ways to do what you're trying to do without the need for a `maybeNamed`.  Let's say that the "meta" tasks are added by the plugin `meta`.

1. In your plugin, you could apply `meta` before any code that requires meta tasks. In that case, the tasks will exist and `named(...)` will do what you want.
2. From your plugin, you could use `pluginManager.withPlugin("meta") {}` to configure the meta tasks if the meta plugin applied. In that callback, you can use `named(...)` and it'll work the way you want.

Both of these are less tricky to get right than a `maybeNamed(...)` that may not work because the meta plugin was never applied or you mistyped the meta task name or the meta task changed names. It also gives Gradle more information.

**Comment by big-guy** [comment]
@Tapchicoma An example of where this would be useful would be helpful.

**Comment by JamesXNelson** [comment]
Apologies on the delay replying; this slipped off my radar during the holiday season.

> A project is intended to be evaluated synchronously, so I don't think this is a concern. Any concurrent evaluation today is a bug.

Even without concurrent execution, the use of closures, callbacks and events (.afterEvaluate(), all(), etc), the execution order of code is deterministic (hashmap ordering notwithstanding), but still asynchronous in nature.  Deferred task realization exacerbates a build engineer's difficulty in (accurately) knowing the exact execution order of all code in all build scripts and plugins.  While it's a good thing to defer or skip work, once you start registering configuration instead of executing it, you generally need to stop checking state and start monitoring / reacting to it.

Asking if there is a task registered _right now_ can be safe if you 100% know that all necessary tasks have already been registered (i.e. you create the task, or a plugin you apply creates it immediately, 100% of the time).  Plugin authors should strive to register everything that is necessary immediately, but even then, you can't (easily) control whether consumer of your code applied plugin-a's callback runs before or after plugin-b's callback (where neither a nor b apply each other).  I was excited about the model/rule based approach Gradle had taken, because I felt it was an elegant solution to such problems: reacting to configuration, rather than "wait just long enough for the right scripts and callbacks to evaluate before checking under the hood of the state machine". 

>> I find that most of the time that I have to realize a task when I don't really want to is to capture dependsOn / finalizedBy semantics

>Could you describe this a bit more?

One example of this is a lifecycle task.  Imagine I create an `initEnv` task, and that _only runtime-related tasks need the environment setup._  I want to create multiple tasks that initEnv both depends on and is finalized by, perhaps with more lifecycle preInitEnv and postInitEnv lifecycle tasks bound around that task.  Between these tasks, there can be mustRunAfter/shouldRunAfter semantics as well.  If I have to realize a task to express these task execution graph mappings, it makes it hard to avoid configuring optional tasks that would not have otherwise needed to exist.

The actual use case here is that external jar dependencies need to be processed before they can be used at runtime, but it's legal and faster to compile against them without any such work being done.  There are sets of tasks added per "group:name:version", and orchestrated around those lifecycle tasks.  

If only the compile classpath is needed, I don't even want to register any of these initEnv tasks.

Since it's relatively expensive to process every transitive dependency per project, it would actually be nice to just `project(':jar-preprocessor').tasks.maybeRegister("g:n:v", {...})`, so every copy of g/n-v.jar is squashed into a single task whose finalized output is a post-processed artifact.  Of course, this is what artifact transforms are supposed to do anyway (what I'm aiming for now), so maybe not the best use case for _this_ API change.

The use case of "task execution graph wiring between un-realized tasks" may still be valid, and suggests that TaskProvider and Task might benefit from a shared superinterface (say, HasTaskWiring) for dependsOn/finalizedBy/shouldRunAfter/mustRunAfter, which can be wholly wired through TaskProvider instances.

**Comment by JamesXNelson** [comment]
I don't like `TaskProvider TaskContainer.maybeNamed()` because it only means "has the task been registered _at the current point in time_"; really, if anything, `boolean TaskContainer.isRegistered(CharSequence)` might be better, though still less safe than `void TaskContainer.ifRegistered(CharSequence, Action)`, since that only exposes the condition of "the task is registered / will be registered during this build", without an (unsafe) else callback.  It's totally cool if you want to know if anybody has registered a task so you can respond to it, but really hard to tell you a task is not registered now, nor _will be_ registered in the future.

Once you start deferring execution a little, you start needing to defer it everywhere, and the more synchronous return values you expose, the more "fun" race conditions you invite.  I really like the way gradle is moving w.r.t. lazy configuration, and look forward to seeing more and more code live inside callbacks that are only invoked if needed. :-)

TL;DR: I vote for TaskContainer.ifRegistered and TaskContainer.maybeRegister instead, as they likely give more power _and_ safety than synchronous alternatives.

**Comment by tasomaniac** [comment]
@JamesXNelson I understand what you mean but isn't it the same for the simple `TaskContainer.named()` method? If the task has **not** been registered at the current point in time, `named` just throws.

**Comment by JamesXNelson** [comment]
I think `named()` is a little different.  You are not checking whether it has been registered or not... you _know_ it has been registered, and expect it to break if not.  It covers the "yes, it has already been registered", not the dangerous / racy "no, it hasn't been registered _yet_" cases.  It doesn't return null or false or something you _should_ respond to; it throws, and if you are really catching that, you are ignoring a strong warning that you should probably be checking some other state beforehand (i.e. "Is the plugin applied?").  When dealing with asynchronous state machines, `getOrThrow` is valid since there are times when you know something must have happened (I applied the JavaPlugin, now I can get the `jar` TaskProvider), but `getOrReturnNull` means you don't know if it _has happened yet_, and now you're sensitive to timing / races (you get null back, assume the task won't be created, but then it is a little later... welcome to complex bug city).

If you call `named()` and catch the result, you are opening Pandora's Shiny Box of Race Conditions.  Maybe you are working in a tightly controlled environment (internal corporate tools) and can verify it's safe to do so, but if you're developing open source plugins to be applied and configured by others, it's probably a really bad idea to catch that exception.  About the only time I could see you wanting something like that would be if you are expecting users to optionally register a task with a given name in their scripts, and then respond to that.  In such a case, you'd likely be better off with an extension that can create the task (instead of magic task naming conventions), so you can just check the extension instead... 

Basically, it's always safe to respond to something that has happened, but it's not safe to respond to something that hasn't happened _yet_.  Having some distinct "checkpoints" during configuration phase can make it possible to deal in these murky waters, but even then, it would be best to avoid it altogether.

Maybe a gradle team member will disagree with me and set me straight here, but I really think that all lazy configuration APIs need to move towards callbacks instead of synchronous return values (i.e. subscribe and respond to state rather than check it at the current point in time).

**Comment by JamesXNelson** [comment]
Hm.  Do you know which plugins are adding the tasks, and are they added immediately?

If so, you can either do `project.plugins.withType(OtherPlugin).all { ...task will exist... }`
or `project.plugins.apply(OtherPlugin)` as many times as you want, to ensure it does exist.

As a worst case scenario, you _could_ use `tasks.configureEach` to get notified that the other task is created, but this might be dangerous; if the other task is also lazily configured, then you (probably) wouldn't get notified in time to mutate the task graph.

Another possible workaround might be for you to always create a do-nothing lifecycle task, and have that depend on a registered "actually does more expensive task configuration" task that you _do_ fully own.  This way, you _can_ use something like `tasks.configureEach` to search for other plugin's tasks, and just register dependsOn to your lifecycle task (which you can setup immediately / before task graph is finalized).

**Comment by tasomaniac** [comment]
Your 2nd option does not work for the exact same reason. The other tools task is also lazy and if I do `configureEach`, it never gets called. 

The last option is interesting. I was doing that already, but I use register for my empty task instead of create. Then inside the configuration of that empty task, Gradle throws and exception since it is too late to configure other tasks. But if I immediately create my empty task, then it might work.

**Comment by big-guy** [comment]
@tasomaniac I'm trying to understand your use case/what you're trying to accomplish.

With this extension:
```
staticAnalysis {
    pmd {
         // Your config
    }
}
```

Does the `pmd` mean "create a PMD task and configure it" or does it mean "here is some configuration for how to create other PMD tasks"?

> Original plugin may have applied and tasks may be created already. Or not.

This makes me think that the `pmd {}` above applies to configuring a single task?

Is the idea that someone would apply your plugin and configure it and get PMD/FindBugs/whatever tasks based on how your plugin configures things?  Or someone could apply the built-in PMD/FindBugs/whatever tasks and _then_ your plugin and be able to configure your extension in a similar way, but you would know how to configure the existing tasks?

> This extension can put and run multiple times. Usually when extension is configured multiple times, you are supposed to only change your extension and then the tasks would pick up the properties from the extension. But here, since we register tasks we don't own, we just use maybeCreate and it works quite well.

I'm not sure what "run multiple times" and "only change your extension" here means. It sounds like you're doing something like:
```
staticAnalysis {
    pmd {
         // Your config
    }
}
// at this point, all PMD tasks have been created based on the configuration above
staticAnalysis {
    findbugs {
         // Your config
    }
}
// at this point, all FindBugs tasks have been created based on the configuration above
```

We've talked about, but we haven't started working on something that would let you connect the configuration of one thing to the configuration of another.  e.g., to configure the task container, Gradle must first configure this other container because the task container derives tasks from it. 

@JamesXNelson I would push back on describing this as a race condition or a concurrency problem.  There's no concurrency involved here. There _is_ an inherent ordering question and where we checkpoint things (like you say).

It seems to me that any sort of `maybe...` or `isRegistered` API is going to make the ordering issue worse, not better.

Reading the comments so far, I sort of see a few use cases where `maybe...` methods look appealing, but could be solved in some other way:

- Deriving one task's configuration from another task, if it exists. I think we want to be able to do this directly somehow vs forcing people to use `maybe...` to simulate it.
- Registering a task that needs to exist in multiple plugins, but neither plugin has the right to "own" the task. I think this is a sign of some sort of intermediate thing missing and the task is being used to represent the intermediate thing and the task that does the work. I think `maybe...` methods just intertwine the two plugins more intead of moving the responsibility to a definite owner.

**Comment by tasomaniac** [comment]
I will give you more detailed answer later. But in short, we support all of what you described. On top of that the number of tasks per tool varies based on the number of  sourceSets. We are currently very flexible and support as much scenario as possible.

**Comment by melix** [comment]
My use case for this is simple: I want to create a `Jar` task if the user hasn't done it already, because when you create your own source set, we don't automatically create a `Jar` task for it, and my plugin needs the jar. So first I need to check if the jar task exists, and if it doesn't, create it.

If I use `findByName`, it would force the realization of the `Jar` task too early.
If I use `named`, it would fail in case the `Jar` task doesn't exist.

So the trick I used is to call `if (!tasks.getNames().contains(...))` and then register a task in this case. This avoids the realization, but it would be nice to have a `maybeNamed` variant indeed.

**Comment by tasomaniac** [comment]
I thought `tasks.getNames()` would cause realization of all tasks. Isn't it that the case?

**Comment by melix** [comment]
As far as I can tell, it doesn't.

**Comment by lacasseio** [comment]
Maybe we should have `named(String) ` return a provider that can return false to `isPresent()` and keep throwing an exception for `named(String, Action)` and `configure(Action)` if the task doesn't exists. What do you guys think?

**Comment by melix** [comment]
I'm in favor of keeping backwards compatibility as much as possible, so I would prefer another method. Especially having the same method name but different behavior based on the number of arguments can be confusing.

**Comment by orangy** [comment]
Applying it to any NDOC, not just tasks, I'd like to have two additional methods on it:

`container.withName("foo") { … }` that would call lambda immediately if `foo` is already realized, or later when realization occurs, or never if `foo` is never registered.

`container.allNames { name -> …}` similar to `all` but not realizing any tasks and just calling when domain object is registered.

**Comment by tasomaniac** [comment]
Awesome ideas. 👍 for

- A method to return TaskProvider immediately from a name. 
- tasks.allNames to get all providers as soon as they're registered

**Comment by tasomaniac** [comment]
`container.allNames` would be awesome because I just want my (already created) task to `dependsOn` other providers. I would do

```
tasks.withType(FooTask).allNames {
    otherTask.dependsOn it
}
```

**Comment by orangy** [comment]
My use case is that I have a container that matches by name another container. 

```
sourceSets.all { sourceSet ->
   mycontainer.withName(sourceSet.name) {
       // (1)
   }
}
```

I want `(1)` to run when both sourceSet is realized and mycontainer domain object is realized. I don't want all other mycontainer element to be realized, as is the case with `all`. I want it to work when reigstrations are in any order:

```
sourceSets {
   bar
}
foo {
    bar {}
    baz {}
}
sourceSets {
   baz
}
```

execution order:
* bar source set is registered and realized
* call to foo.withName("bar") registers a by-name-listener for foo
* foo is configured
* bar is registered. Listener above is triggered, causing bar to be realized. 
* baz is registered, no realization 
* baz source set is registered and realized
* call to foo.withName("baz") completes synchronously, because baz is already registered, so it is realized at this point and the lambda is called

**Comment by vlsi** [comment]
Here's `maybeRegister` implemented by Gradle grandmasters: https://github.com/gradle/gradle/blob/5ec3f672ed600a86280be490395d70b7bc634862/subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmVariantBuilder.java#L275-L276

I would steal it from there :)

**Comment by tasomaniac** [comment]
So, the only way to check a task with a name exists without forcing it to be configured is to check if it is in `tasks.getNames()`. It is weird that this is not provided as an API.

**Comment by melix** [comment]
Gradle Bot, I beg you don't close this one!

**Comment by DavidGregory084** [comment]
Another use case for this one  - configuring code generation tasks so that they don't cause implicit task dependency problems.

See e.g. [this code](https://github.com/opencastsoftware/gradle-build-info/blob/main/src/main/java/com/opencastsoftware/gradle/buildinfo/BuildInfoPlugin.java#L116), where I want to register a dependency between `sourcesJar`, which is not always registered, and my plugin's task, without Gradle either blowing up with `TaskNotFoundException` or just not doing the configuration and still complaining about implicit task dependencies later on.

**Comment by rkrisztian** [comment]
> `container.allNames` would be awesome because I just want my (already created) task to `dependsOn` other providers. I would do
> 
> ```
> tasks.withType(FooTask).allNames {
>     otherTask.dependsOn it
> }
> ```

Isn't that the same as the following?

```groovy
otherTask.dependsOn tasks.withType(FooTask)
```

**Comment by Goooler** [comment]
Introducing `maybeRegister` should be great, which will also address:

- #6243
- #8106
